### PR TITLE
OPML generation for blogs list

### LINKED ISSUE #289: Blogs as OPML import
I don't say it nearly enough, but I reference this project a *ton* -- thank you @daveaglick!

I had an idea and if you're up for it, let me know and I can submit a PR.

You have a number of great blog resources that I'd love to periodically bring into my Feedly account, which allows the import of OPML. Any objection to me adding the generation of an OPML file and optionally a link to the file itself for download? 

If you're OK with the idea, let me know where you'd like to the file to reside and where (if anywhere) you'd like me to add a link to it on the site, and I'll submit the PR. (if no preferences, let me know and I can take a first pass at it based on my instincts).

### PR DESCRIPTION
**Pull request by SeanKilleen**
Resolves #289. 

### High-Level Changes

* Adds OPML generation pipeline
* Adds VS Code workplace setting to treat `.wyam` like C#. (let me know if you want this to be a separate PR)
* Adds some `README` clarifications around token generation.

### Work / Remaining

* [x] Create Layout file for OPML
* [x] Create Pipeline for OPML generation
* [x] Figure out how to get YAML contents into a list of objects in wyam
* [x] Figure out how to use a collection of elements in Razor via wyam
* [x] Get the YAML contents into a list in the  pipeline
* [x] Use the razor layout in the pipeline.
* [x] Save the file
* [x] Add a link to the file
* [x] Get review / input
* [ ] Squash the PR

### CODE REVIEW

**Comment by SeanKilleen** [comment]
@daveaglick quick wyam question -- what's the appropriate strategy for getting my `IEnumerable<IDocument>` with the appropriate metadata into a Razor page? 

* Would I need to create my own list of metadata items and then set the model to `IEnumerable<MetadataItems>`?
* Is there a convention for a Razor page that's being passed an enumerable that it loops over?
* Something else I'm missing?

File: README.md [hunk 1]
```diff
@@ -6,10 +6,8 @@
 
 To build the site for previewing, you'll need to set the following environment variables:
 
-```
-DISCOVERDOTNET_GITHUB_TOKEN
-DISCOVERDOTNET_MEETUP_TOKEN
-```
+* `DISCOVERDOTNET_GITHUB_TOKEN`, which can be obtained via creating a [personal access token](https://github.com/settings/tokens) with `public_repo` permissions.
+* `DISCOVERDOTNET_MEETUP_TOKEN`, which can be obtained by visiting [the Meetup API key page](https://secure.meetup.com/meetup_api/key/).
```
Comment (daveaglick) [inline README.md]: Very nice, thanks for the clarification - hopefully you won't be the last to work on the site and this will help those that follow :smile:

File: config.wyam [hunk 2]
```diff
@@ -226,6 +226,13 @@ Pipelines.Add("Post Search Data",
     WriteFiles((doc, ctx) => "data/search/posts.json")
 );
 
+Pipelines.Add("OPML feed export", 
+    ReadFiles("data/blogs/*.yml"),
```
Comment (daveaglick) [inline config.wyam]: You probably don't need to read the files and process them again since they alreayd get read and are available in the previous "Blog" pipeline. You should be able to replace the `ReadFiles`, `Yaml`, and `Content` modules here with a single `Documents("Blogs")`:

```
Pipelines.Add("OPML",
    Documents("Blogs"),
    Execute((inputs, ctx, _) => GetOpmlData(inputs, ctx)),
);
```

**Comment by SeanKilleen** [comment]
@daveaglick I think this is ready for review if/when you get the chance.

I'd love if you could please confirm that all the blogs output to the OPML in the build process (or provide a quick ability for me to do so). It seems like on the preview build only 5 will be outputted, but when running the build with no target there also only seemed to be 5 outputted and i want to make sure I didn't miss something there.

Once this gets your 👍 I'll squash it.

File: .vscode/settings.json [hunk 3]
```diff
@@ -0,0 +1,5 @@
+{
+    "files.associations": {
+        "*.wyam": "csharp"
```
Comment (SeanKilleen) [inline .vscode/settings.json]: ℹ️ FYI this is to ensure that anyone who opens this in VSCode will have the `.wyam` files mapped to C# in the editor for highlighting, etc. (can be a separate PR if you want)

File: config.wyam [hunk 4]
```diff
@@ -148,6 +148,7 @@ Pipelines.Add("Blogs",
         Take(5)),
     Yaml(),    
     Content(string.Empty),
+    Execute((inputs, ctx, _) => GetOpmlData(inputs, ctx)),
```
Comment (daveaglick) [inline config.wyam]: I think we can simplify this even further - you may not even need `GetOpmlData()` now that you're relying on metadata from the "Blogs" pipeline in the `feeds.cshtml` file. It looks like the `GetOpmlData()` method is mostly responsible for setting a "Title" metadata value when one doesn't exist. That's actually on of the jobs of the `GetFeedData()` method right below it - and that method sets the "Title" from the actual RSS feed. In fact, the current ordering will actually prevent `GetFeedData()` from setting the title from the RSS because it only does so if an explicit title isn't set and `GetOpmlData()` right before it sets the "Title" metadata in that case (so `GetFeedData()` would never get a blank title that signals the RSS one should be used). At a minimum, `GetOpmlData()` should be moved below `GetFeedData()` so it only sets the backup title if one wasn't provided in the YAML _and_ couldn't be obtained from the RSS feed.

File: config.wyam [hunk 5]
```diff
@@ -845,6 +852,32 @@ public class FeedItemData
...
+
+                title = feed.Substring(from, to - from);
+            }
+
+            MetadataItems metadata = new MetadataItems(){
+                { "Title", title },
+                { "Feed", feed }
+            };
+
+            return context.GetDocument(input, metadata);
```
Comment (daveaglick) [inline config.wyam]: Creating a new document is a little expensive (the engine caches documents on creation - one of the reasons we have to use a factory to get them). Can we move this inside the `if (string.IsNullOrWhiteSpace(title))` and return the original document (`return input`) if the title isn't changed?

File: config.wyam [hunk 6]
```diff
@@ -845,6 +852,32 @@ public class FeedItemData
...
+                
+                int from = feed.IndexOf(slashes) + slashes.Length;
+                int to = feed.IndexOf("/", from);
+
+                title = feed.Substring(from, to - from);
+            }
+
+            MetadataItems metadata = new MetadataItems(){
+                { "Title", title },
+                { "Feed", feed }
```
Comment (daveaglick) [inline config.wyam]: Since "Feed" isn't being given a new value you can omit it from the `MetadataItems` - the existing value from the `input` document will "fall through" and still be available in the new document.

File: config.wyam [hunk 7]
```diff
@@ -226,6 +227,12 @@ Pipelines.Add("Post Search Data",
     WriteFiles((doc, ctx) => "data/search/posts.json")
 );
 
+Pipelines.Add("OPML feed export", 
+    ReadFiles("feeds.cshtml"),
+    Razor(),
+    WriteFiles(".opml")
+);
```
Comment (daveaglick) [inline config.wyam]: The `feeds.cshtml` file should probably be named `_feeds.cshtml`, otherwise the "Pages" pipeline will also read it and might overwrite it (or this pipeline would overwrite that one - either way not ideal). The "Pages" pipeline reads all `*.cshtml` files unless they start with an underscore. If you do that, then this pipeline should probably be:
```
Pipelines.Add("OPML feed export", 
    ReadFiles("_feeds.cshtml"),
    Razor(),
    WriteFiles((doc, ctx) => "feeds.opml")
);
```

**Comment by SeanKilleen** [comment]
@daveaglick this was all super helpful! Definitely helped unstick me on some things. Should have it wrapped up shortly, and I'll get rid of the OPML method in favor of a modification to the `GetFeedData` method. Thanks for all the pointers / patience!

**Comment by SeanKilleen** [comment]
@daveaglick I pushed a bunch of updates, looks like we're just stuck on one last item.

Using the suggested rename to `_feeds.html`, it looks like the opml file no longer generates, and I see in the output:

```output
     Executing pipeline "OPML feed export" (13/35) with 3 child module(s)
             Ignoring 1 inputs due to source file name prefix
         Executed pipeline "OPML feed export" (13/35) in 10 ms resulting in 0 output document(s)
```

I'll keep looking into it, but I think if we can fix the fact that the input is being ignored to the prefix, the (much simplified, thanks!) PR should be ready to squash & go shortly.

**Comment by SeanKilleen** [comment]
Looks like ignoring files prefixed with an underscore is part of Wyam's Razor module, so we may need to rename the file: https://github.com/Wyamio/Wyam/blob/develop/src/extensions/Wyam.Razor/Razor.cs#L180

So we may need  to rename the cshtml file.

**Comment by SeanKilleen** [comment]
Last piece, @daveaglick -- I'm not sure if it's because of a build command that I'm not familiar with, but it appears that no matter what, the `opml` file generated only consists of 4 entries.

If I run `.\build`, does that automatically imply `-target preview`? I want to make sure the full file can be generated for all of the feeds..

**Comment by daveaglick** [comment]
Correct, the default target is set to `Preview`, which limits data to 5 items each (the full build takes a while to run):
https://github.com/daveaglick/discoverdotnet/blob/2c94428259338521343296d79767662f011e5a0c/build.cake#L275-L276

To build it with everything, run `build -target build`...and then go do something else for a while :) I've confirmed that all the blogs are in the OPML feed when running the full build and I was even able to import it into Feedly.

This is awesome, thanks! I'll go ahead and merge it in now.
